File: jlox/src/main/java/fr/ocus/lox/jlox/Interpreter.java
package fr.ocus.lox.jlox;

import java.io.PrintStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author Matthieu Honel <ocus51@gmail.com>
 * @since 2017-08-03
 */
public class Interpreter implements Expr.Visitor<Object>, Stmt.Visitor<Void> {
    final Environment globals = new Environment();
    private Environment environment = globals;
    private final Map<Expr, Integer> locals = new HashMap<>();
    private final PrintStream printStream;
    private final PrintStream errorStream;

    Interpreter() {
        this(System.out, System.err);
    }

    Interpreter(PrintStream printStream, PrintStream errorStream) {
        this.printStream = printStream;
        this.errorStream = errorStream;
        globals.define("clock", new LoxCallable() {
            @Override
            public String toString() {
                return "<fn:lox clock>";
            }

            @Override
            public int arity() {
                return 0;
            }

            @Override
            public Object call(Interpreter interpreter, List<Object> arguments) {
                return (double) System.currentTimeMillis() / 1000.0;
            }
        });
        globals.define("dump_env", new LoxCallable() {
            @Override
            public String toString() {
                return "<fn:lox dump_env>";
            }

            @Override
            public int arity() {
                return 0;
            }

            @Override
            public Object call(Interpreter interpreter, List<Object> arguments) {
                Map<String, Object> values = interpreter.environment.getAllValues();
                print("Env dump start: (", values.size(), ")\n");
                for (String name : values.keySet()) {
                    print("- ", name, " = ", values.get(name), "\n");
                }
                print("Env dump end\n");
                return null;
            }
        });
    }

    void interpret(List<Stmt> statements) {
        try {
            for (Stmt statement : statements) {
                execute(statement);
            }
        } catch (StackOverflowError error) {
            JLox.runtimeError(errorStream, new RuntimeError(new Token(TokenType.EOF, "", "", -1), "Stack overflow."));
        } catch (RuntimeError error) {
            JLox.runtimeError(errorStream, error);
        }
    }

    @Override
    public Object visitBinaryExpr(Expr.Binary expr) {
        Object left = evaluate(expr.left);
        Object right = evaluate(expr.right);

        switch (expr.operator.type) {
            case MINUS:
                checkNumberOperands(expr.operator, left, right);
                return (double) left - (double) right;
            case PLUS:
                if (left instanceof Double && right instanceof Double) {
                    return (double) left + (double) right;
                }

                if (left instanceof String && right instanceof String) {
                    return (String) left + right;
                }

                throw new RuntimeError(expr.operator, "Operands must be two numbers or two strings.");
            case SLASH:
                checkNumberOperands(expr.operator, left, right);
                checkNotZeroOperand(expr.operator, right);
                return (double) left / (double) right;
            case STAR:
                checkNumberOperands(expr.operator, left, right);
                return (double) left * (double) right;
            case GREATER:
                checkNumberOperands(expr.operator, left, right);
                return (double) left > (double) right;
            case GREATER_EQUAL:
                checkNumberOperands(expr.operator, left, right);
                return (double) left >= (double) right;
            case LESS:
                checkNumberOperands(expr.operator, left, right);
                return (double) left < (double) right;
            case LESS_EQUAL:
                checkNumberOperands(expr.operator, left, right);
                return (double) left <= (double) right;
            case BANG_EQUAL:
                return !isEqual(left, right);
            case EQUAL_EQUAL:
                return isEqual(left, right);
        }

        // Unreachable.
        return null;
    }

    @Override
    public Void visitFunctionStmt(Stmt.Function stmt) {
        LoxFunction function = new LoxFunction(stmt, environment, false);
        environment.define(stmt.name.lexeme, function);
        return null;
    }

    @Override
    public Void visitReturnStmt(Stmt.Return stmt) {
        Object value = null;
        if (stmt.value != null) value = evaluate(stmt.value);

        throw new Return(value);
    }

    @Override
    public Object visitCallExpr(Expr.Call expr) {
        Object callee = evaluate(expr.callee);

        List<Object> arguments = new ArrayList<>();
        for (Expr argument : expr.arguments) {
            arguments.add(evaluate(argument));
        }

        if (!(callee instanceof LoxCallable)) {
            throw new RuntimeError(expr.paren, "Can only call functions and classes.");
        }

        LoxCallable function = (LoxCallable) callee;
        if (arguments.size() != function.arity()) {
            throw new RuntimeError(expr.paren, "Expected " + function.arity() + " arguments but got " + arguments.size() + ".");
        }
        return function.call(this, arguments);
    }

    @Override
    public Object visitGetExpr(Expr.Get expr) {
        Object object = evaluate(expr.object);
        if (object instanceof LoxInstance) {
            return ((LoxInstance) object).getProperty(expr.name);
        }

        throw new RuntimeError(expr.name, "Only instances have properties.");
    }

    @Override
    public Object visitGroupingExpr(Expr.Grouping expr) {
        return evaluate(expr.expression);
    }

    @Override
    public Object visitIndexGetExpr(Expr.IndexGet expr) {
        Object index = expr.index.accept(this);
        if (!(index instanceof Double)) {
            throw new RuntimeException(); // not a number
        }
        double doubleIndex = (double) index;
        if ((doubleIndex != Math.floor(doubleIndex)) || Double.isInfinite(doubleIndex)) {
            throw new RuntimeException(); // not a whole number
        }
        Object object = expr.object.accept(this);
        if (object instanceof List) {
            return ((List) object).get((int) doubleIndex);
        }
        else if (object instanceof String) {
            return ((String) object).charAt((int) doubleIndex);
        }
        throw new RuntimeException(); // not a list
    }

    @Override
    public Object visitIndexSetExpr(Expr.IndexSet expr) {
        Object index = expr.index.accept(this);
        if (!(index instanceof Double)) {
            throw new RuntimeException(); // not a number
        }
        double doubleIndex = (double) index;
        if ((doubleIndex != Math.floor(doubleIndex)) || Double.isInfinite(doubleIndex)) {
            throw new RuntimeException(); // not a whole number
        }
        Object object = expr.object.accept(this);
        if (object instanceof List) {
            ((List) object).set((int) doubleIndex, expr.value.accept(this));
        }
        else if (object instanceof String) {
            throw new RuntimeException(); // hmmm strings are immutable
        }
        else {
            throw new RuntimeException(); // not a list nor a string
        }
        return null;
    }

    @Override
    public Object visitLogicalExpr(Expr.Logical expr) {
        Object left = evaluate(expr.left);

        if (expr.operator.type == TokenType.OR) {
            if (isTruthy(left)) return left;
        } else {
            if (!isTruthy(left)) return left;
        }

        return evaluate(expr.right);
    }

    @Override
    public Object visitSetExpr(Expr.Set expr) {
        Object value = evaluate(expr.value);
        Object object = evaluate(expr.object);

        if (object instanceof LoxInstance) {
            ((LoxInstance) object).fields.put(expr.name.lexeme, value);
            return value;
        }

        throw new RuntimeError(expr.name, "Only instances have fields.");
    }

    @Override
    public Object visitThisExpr(Expr.This expr) {
        return environment.get(expr.keyword);
    }

    @Override
    public Object visitSuperExpr(Expr.Super expr) {
        int distance = locals.get(expr);
        LoxClass superclass = (LoxClass) environment.getAt(distance, "super");
        LoxInstance receiver = (LoxInstance) environment.getAt(distance - 1, "this");
        LoxFunction method = superclass.findMethod(receiver, expr.method.lexeme);
        if (method == null) {
            throw new RuntimeError(expr.method, "Undefined property '" + expr.method.lexeme + "'.");
        }
        return method;
    }

    @Override
    public Object visitLiteralExpr(Expr.Literal expr) {
        return expr.value;
    }

    @Override
    public Object visitUnaryExpr(Expr.Unary expr) {
        Object right = evaluate(expr.right);

        switch (expr.operator.type) {
            case BANG:
                return !isTruthy(right);
            case MINUS:
                checkNumberOperand(expr.operator, right);
                return -(double) right;
        }

        // Unreachable.
        return null;
    }

    @Override
    public Object visitVariableExpr(Expr.Variable expr) {
        return lookUpVariable(expr.name, expr);
    }

    @Override
    public Object visitArrayExpr(Expr.Array expr) {
        List<Object> array = new ArrayList<>(expr.elements.size());
        for (Expr element : expr.elements) {
            array.add(element.accept(this));
        }
        return array;
    }

    @Override
    public Object visitAssignExpr(Expr.Assign expr) {
        Object value = evaluate(expr.value);

        Integer distance = locals.get(expr);
        if (distance != null) {
            environment.assignAt(distance, expr.name, value);
        } else {
            environment.assign(expr.name, value);
        }
        return value;
    }

    @Override
    public Void visitExpressionStmt(Stmt.Expression stmt) {
        evaluate(stmt.expression);
        return null;
    }

    @Override
    public Void visitIfStmt(Stmt.If stmt) {
        if (isTruthy(evaluate(stmt.condition))) {
            execute(stmt.thenBranch);
        } else if (stmt.elseBranch != null) {
            execute(stmt.elseBranch);
        }
        return null;
    }

    @Override
    public Void visitPrintStmt(Stmt.Print stmt) {
        Object value = evaluate(stmt.expression);
        print(value, "\n"); // platform-independent newline
        return null;
    }

    private void print(Object... values) {
        for (Object value : values) {
            printStream.print(stringify(value));
        }
    }

    @Override
    public Void visitWhileStmt(Stmt.While stmt) {
        while (isTruthy(evaluate(stmt.condition))) {
            execute(stmt.body);
        }
        return null;
    }

    @Override
    public Void visitVarStmt(Stmt.Var stmt) {
        Object value = null;
        if (stmt.initializer != null) {
            value = evaluate(stmt.initializer);
        }

        environment.define(stmt.name.lexeme, value);
        return null;
    }

    @Override
    public Void visitBlockStmt(Stmt.Block stmt) {
        executeBlock(stmt.statements, new Environment(environment));
        return null;
    }

    @Override
    public Void visitClassStmt(Stmt.Class stmt) {
        environment.define(stmt.name.lexeme, null);

        Object superclass = null;
        if (stmt.superclass != null) {
            superclass = evaluate(stmt.superclass);
            if (!(superclass instanceof LoxClass)) {
                throw new RuntimeError(stmt.name, "Superclass must be a class.");
            }

            environment = new Environment(environment);
            environment.define("super", superclass);
        }
        Map<String, LoxFunction> methods = new HashMap<>(stmt.methods.size());
        for (Stmt.Function method : stmt.methods) {
            LoxFunction function = new LoxFunction(method, environment, method.name.lexeme.equals("init"));
            methods.put(method.name.lexeme, function);
        }
        LoxClass cls = new LoxClass(stmt.name.lexeme, (LoxClass) superclass, methods);
        if (superclass != null) {
            environment = environment.enclosing;
        }

        environment.assign(stmt.name, cls);
        return null;
    }

    private void execute(Stmt stmt) {
        stmt.accept(this);
    }

    void executeBlock(List<Stmt> statements, Environment environment) {
        Environment previous = this.environment;
        try {
            this.environment = environment;

            for (Stmt statement : statements) {
                execute(statement);
            }
        } finally {
            this.environment = previous;
        }
    }

    void resolve(Expr expr, int depth) {
        locals.put(expr, depth);
    }

    private Object evaluate(Expr expr) {
        return expr.accept(this);
    }

    private Object lookUpVariable(Token name, Expr expr) {
        Integer distance = locals.get(expr);
        if (distance != null) {
            return environment.getAt(distance, name.lexeme);
        }
        return globals.get(name);
    }

    private void checkNumberOperand(Token operator, Object operand) {
        if (operand instanceof Double) return;
        throw new RuntimeError(operator, "Operand must be a number.");
    }

    private void checkNumberOperands(Token operator, Object left, Object right) {
        if (left instanceof Double && right instanceof Double) return;
        throw new RuntimeError(operator, "Operands must be numbers.");
    }

    private void checkNotZeroOperand(Token operator, Object operand) {
        if ((double) operand != 0) return;
        throw new RuntimeError(operator, "Division by zero.");
    }

    private boolean isTruthy(Object object) {
        return object != null && (!(object instanceof Boolean) || (boolean) object);
    }

    private boolean isEqual(Object a, Object b) {
        // nil is only equal to nil.
        return a == null && b == null || a != null && a.equals(b);
    }

    private String stringify(Object object) {
        if (object == null) return "nil";

        // Hack. Work around Java adding ".0" to integer-valued doubles.
        if (object instanceof Double) {
            String text = object.toString();
            if (text.endsWith(".0")) {
                text = text.substring(0, text.length() - 2);
            }
            return text;
        }

        return object.toString();
    }

    private void niy(Expr expr) {
        throw new RuntimeException("Expr: not implemented yet " + expr.getClass().getName() + ".");
    }

    private void niy(Stmt stmt) {
        throw new RuntimeException("Stmt: not implemented yet " + stmt.getClass().getName() + ".");
    }
}

File: pylox/pylox/interpreter.py
import sys
import time
from typing import Any, Dict, List

from .callable import LoxCallable, LoxClass, LoxFunction, LoxInstance, Return
from .environment import Environment
from .error import PyLoxRuntimeError
from .expr import AbstractExpr, ExprAssign, ExprBinary, ExprCall, ExprGet, ExprGrouping, ExprLiteral, ExprLogical, \
    ExprSet, ExprSuper, ExprThis, ExprUnary, ExprVariable
from .interface import EnvironmentInterface, InterpreterInterface
from .stmt import AbstractStmt, StmtBlock, StmtClass, StmtExpression, StmtFunction, StmtIf, StmtPrint, StmtReturn, \
    StmtVar, StmtWhile
from .token import Token, TokenType


class StdClock(LoxCallable):
    def call(self, interpreter: InterpreterInterface, arguments: List[Any]):
        return time.time()

    def arity(self) -> int:
        return 0

    def __str__(self):
        return '<fn:lox clock>'


class StdGetEnv(LoxCallable):
    def __init__(self):
        super(StdGetEnv, self).__init__()

    def call(self, interpreter: InterpreterInterface, arguments: List[Any]):
        return interpreter.get_env()

    def arity(self) -> int:
        return 0

    def __str__(self):
        return '<fn:lox get_env>'


class Interpreter(InterpreterInterface, AbstractStmt.Visitor, AbstractExpr.Visitor):
    def __init__(self, out_file=sys.stdout, error_file=sys.stderr):
        self._out_file = out_file
        self._error_file = error_file
        self.globals = Environment()
        self._environment = self.globals
        self._locals = {}  # type: Dict[AbstractExpr, int]

        self.globals.define('clock', StdClock())
        self.globals.define('get_env', StdGetEnv())

    def get_env(self):
        return self._environment

    def interpret(self, statements: List[AbstractStmt]):
        try:
            for statement in statements:
                self._execute(statement=statement)
        except PyLoxRuntimeError as error:
            from . import PyLox
            PyLox.runtime_error(error=error, error_file=self._error_file)

    def execute_block(self, statements: List[AbstractStmt], environment: EnvironmentInterface):
        previous = self._environment
        try:
            self._environment = environment

            for statement in statements:
                self._execute(statement=statement)
        finally:
            self._environment = previous

    def resolve(self, expr: AbstractExpr, depth: int):
        self._locals[expr] = depth

    def _execute(self, statement: AbstractStmt):
        statement.accept(visitor=self)

    def _evaluate(self, expr: AbstractExpr):
        return expr.accept(visitor=self)

    def _look_up_variable(self, token: Token, expr: AbstractExpr):
        distance = self._locals.get(expr)
        if distance is not None:
            return self._environment.get_at(distance=distance, name=token.lexeme)
        return self.globals.get(token=token)

    def visit_block_stmt(self, stmt: StmtBlock):
        self.execute_block(statements=stmt.statements, environment=Environment(enclosing=self._environment))
        return None

    def visit_class_stmt(self, stmt: StmtClass):
        self._environment.define(name=stmt.name.lexeme, value=None)

        superclass = None
        if stmt.superclass:
            superclass = self._evaluate(stmt.superclass)
            if not isinstance(superclass, LoxClass):
                raise PyLoxRuntimeError(stmt.name, 'Superclass must be a class.')

            self._environment = Environment(enclosing=self._environment)
            self._environment.define(name='super', value=superclass)

        methods = {}  # type: Dict[str, LoxFunction]
        for method in stmt.methods:
            function = LoxFunction(declaration=method,
                                   closure=self._environment,
                                   is_initializer=method.name.lexeme == 'init')
            methods[method.name.lexeme] = function

        cls = LoxClass(name=stmt.name.lexeme, superclass=superclass, methods=methods)
        if superclass:
            self._environment = self._environment.enclosing

        self._environment.assign(stmt.name, cls)
        return None

    def visit_expression_stmt(self, stmt: StmtExpression):
        return self._evaluate(expr=stmt.expression)

    def visit_function_stmt(self, stmt: StmtFunction):
        function = LoxFunction(declaration=stmt, closure=self._environment, is_initializer=False)
        self._environment.define(name=stmt.name.lexeme, value=function)
        return None

    def visit_if_stmt(self, stmt: StmtIf):
        if Interpreter._is_truthy(value=self._evaluate(expr=stmt.condition)):
            self._execute(statement=stmt.then_branch)
        elif stmt.else_branch:
            self._execute(statement=stmt.else_branch)

    def visit_print_stmt(self, stmt: StmtPrint):
        value = self._evaluate(expr=stmt.expression)
        print(self._stringify(obj=value), end='\n', file=self._out_file)

    def visit_var_stmt(self, stmt: StmtVar):
        value = None
        if stmt.initializer:
            value = self._evaluate(expr=stmt.initializer)

        self._environment.define(name=stmt.name.lexeme, value=value)
        return None

    def visit_return_stmt(self, stmt: StmtReturn):
        value = None
        if stmt.value:
            value = self._evaluate(expr=stmt.value)

        raise Return(value=value)

    def visit_while_stmt(self, stmt: StmtWhile):
        while self._is_truthy(value=self._evaluate(expr=stmt.condition)):
            self._execute(statement=stmt.body)
        return None

    def visit_assign_expr(self, expr: ExprAssign):
        value = self._evaluate(expr=expr.value)

        distance = self._locals[expr] if expr in self._locals else None
        if distance is not None:
            self._environment.assign_at(distance=distance, token=expr.name, value=value)
        else:
            self.globals.assign(token=expr.name, value=value)
        return value

    def visit_binary_expr(self, expr: ExprBinary):
        left = self._evaluate(expr=expr.left)
        right = self._evaluate(expr=expr.right)

        if TokenType.MINUS == expr.operator.type:
            Interpreter._check_number_operands(operator=expr.operator, left=left, right=right)
            return left - right
        if TokenType.PLUS == expr.operator.type:
            left_is_float = Interpreter._is_float(value=left)
            right_is_float = Interpreter._is_float(value=right)
            if left_is_float and right_is_float:
                return left + right
            left_is_str = isinstance(left, str)
            right_is_str = isinstance(right, str)
            if left_is_str and right_is_str:
                return left + right
            if left_is_float and right_is_str:
                return Interpreter._stringify(obj=left) + right
            if left_is_str and right_is_float:
                return left + Interpreter._stringify(obj=right)
            raise PyLoxRuntimeError(expr.operator, message='Operands must be two numbers or two strings.')
        if TokenType.SLASH == expr.operator.type:
            Interpreter._check_number_operands(operator=expr.operator, left=left, right=right)
            Interpreter._check_not_zero(operator=expr.operator, operand=right)
            return left / right
        if TokenType.STAR == expr.operator.type:
            Interpreter._check_number_operands(operator=expr.operator, left=left, right=right)
            return left * right
        if TokenType.GREATER == expr.operator.type:
            Interpreter._check_number_operands(operator=expr.operator, left=left, right=right)
            return left > right
        if TokenType.GREATER_EQUAL == expr.operator.type:
            Interpreter._check_number_operands(operator=expr.operator, left=left, right=right)
            return left >= right
        if TokenType.LESS == expr.operator.type:
            Interpreter._check_number_operands(operator=expr.operator, left=left, right=right)
            return left < right
        if TokenType.LESS_EQUAL == expr.operator.type:
            Interpreter._check_number_operands(operator=expr.operator, left=left, right=right)
            return left <= right
        if TokenType.BANG_EQUAL == expr.operator.type:
            return not Interpreter._is_equal(a=left, b=right)
        if TokenType.EQUAL_EQUAL == expr.operator.type:
            return Interpreter._is_equal(a=left, b=right)

        raise PyLoxRuntimeError(token=expr.operator, message='Unknown binary operator.')

    def visit_call_expr(self, expr: ExprCall):
        callee = self._evaluate(expr=expr.callee)

        arguments = []
        for argument in expr.arguments:
            arguments.append(self._evaluate(expr=argument))

        if not isinstance(callee, LoxCallable):
            raise PyLoxRuntimeError(token=expr.paren, message='Can only call functions and classes.')

        if len(arguments) != callee.arity():
            raise PyLoxRuntimeError(token=expr.paren,
                                    message='Expected ' + str(callee.arity()) + ' arguments but got ' +
                                            str(len(arguments)) + '.')
        return callee.call(interpreter=self, arguments=arguments)

    def visit_get_expr(self, expr: ExprGet):
        object = self._evaluate(expr=expr.object)  # type: LoxInstance
        if isinstance(object, LoxInstance):
            return object.get_property(token=expr.name)

        raise PyLoxRuntimeError(token=expr.name, message='Only instances have properties.')

    def visit_grouping_expr(self, expr: ExprGrouping):
        return self._evaluate(expr=expr.expression)

    def visit_literal_expr(self, expr: ExprLiteral):
        return expr.value

    def visit_logical_expr(self, expr: ExprLogical):
        left = self._evaluate(expr=expr.left)

        if TokenType.OR == expr.operator.type:
            if Interpreter._is_truthy(value=left):
                return left
        else:
            if not Interpreter._is_truthy(value=left):
                return left

        return self._evaluate(expr=expr.right)

    def visit_set_expr(self, expr: ExprSet):
        value = self._evaluate(expr=expr.value)
        object = self._evaluate(expr=expr.object)

        if isinstance(object, LoxInstance):
            instance = object  # type: LoxInstance
            instance.fields[expr.name.lexeme] = value
            return value

        raise PyLoxRuntimeError(token=expr.name, message='Only instances have fields.')

    def visit_this_expr(self, expr: ExprThis):
        return self._environment.get(token=expr.keyword)

    def visit_super_expr(self, expr: ExprSuper):
        distance = self._locals[expr]
        superclass = self._environment.get_at(distance=distance, name='super')  # type: LoxClass
        receiver = self._environment.get_at(distance=distance - 1, name='this')
        method = superclass.find_method(instance=receiver, name=expr.method.lexeme)
        if not method:
            raise PyLoxRuntimeError(token=expr.method, message='Undefined property \'' + expr.method.lexeme + '\'.')
        return method

    def visit_unary_expr(self, expr: ExprUnary):
        right = self._evaluate(expr=expr.right)

        if TokenType.BANG == expr.operator.type:
            return not Interpreter._is_truthy(value=right)
        if TokenType.MINUS == expr.operator.type:
            Interpreter._check_number_operand(operator=expr.operator, operand=right)
            return -right

        raise PyLoxRuntimeError(token=expr.operator, message='Unknown unary operator.')

    def visit_variable_expr(self, expr: ExprVariable):
        return self._look_up_variable(token=expr.name, expr=expr)

    @staticmethod
    def _is_truthy(value: Any):
        if value is None:
            return False
        if isinstance(value, bool):
            return value
        return True

    @staticmethod
    def _is_equal(a: Any, b: Any):
        return type(a) == type(b) and a == b

    @staticmethod
    def _is_float(value: Any):
        return isinstance(value, float)

    @staticmethod
    def _check_number_operand(operator: Token, operand: Any):
        if Interpreter._is_float(value=operand):
            return True
        raise PyLoxRuntimeError(token=operator, message='Operand must be a number.')

    @staticmethod
    def _check_number_operands(operator: Token, left: Any, right: Any):
        if Interpreter._is_float(value=left) and Interpreter._is_float(value=right):
            return True
        raise PyLoxRuntimeError(token=operator, message='Operands must be numbers.')

    @staticmethod
    def _check_not_zero(operator: Token, operand: Any):
        if operand != 0:
            return
        raise PyLoxRuntimeError(token=operator, message='Division by zero.')

    @staticmethod
    def _stringify(obj: Any):
        if obj is None:
            return 'nil'

        if isinstance(obj, bool):
            return 'true' if obj else 'false'

        # HACK remove the ending .0 if necessary
        if Interpreter._is_float(value=obj):
            text = str(obj)
            if text.endswith('.0'):
                text = text[:text.index('.0')]
            return text

        return str(obj)
